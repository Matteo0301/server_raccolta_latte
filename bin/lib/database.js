"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteUser = exports.updateUser = exports.getUsers = exports.getUser = exports.addUser = exports.db = exports.close = exports.connect = void 0;
const mongodb_1 = require("mongodb");
const logger_1 = __importDefault(require("./logger"));
const bcrypt_1 = __importDefault(require("bcrypt"));
let db;
exports.db = db;
let client;
function connect(CONNECTION_STRING, DATABASE) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            client = new mongodb_1.MongoClient(CONNECTION_STRING);
            yield client.connect();
            exports.db = db = client.db(DATABASE);
        }
        catch (error) {
            logger_1.default.error("Error connecting to MongoDB: " + error);
            process.exit(1);
        }
    });
}
exports.connect = connect;
function close() {
    return __awaiter(this, void 0, void 0, function* () {
        yield client.close();
    });
}
exports.close = close;
function generateHash(password) {
    let salt_rounds = 10;
    if (process.env.SALT_ROUNDS) {
        salt_rounds = parseInt(process.env.SALT_ROUNDS);
    }
    return bcrypt_1.default.hashSync(password, bcrypt_1.default.genSaltSync(salt_rounds));
}
function addUser(username, password, admin) {
    return __awaiter(this, void 0, void 0, function* () {
        const hashedPassword = generateHash(password);
        yield db.collection('utenti').insertOne({ username: username, password: hashedPassword, admin: admin });
    });
}
exports.addUser = addUser;
function getUser(username) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield db.collection('utenti').findOne({ username: username });
    });
}
exports.getUser = getUser;
function getUsers() {
    return __awaiter(this, void 0, void 0, function* () {
        return yield db.collection('utenti').find().toArray();
    });
}
exports.getUsers = getUsers;
function updateUser(username, password, admin) {
    return __awaiter(this, void 0, void 0, function* () {
        const hashedPassword = generateHash(password);
        yield db.collection('utenti').updateOne({ username: username }, { $set: { password: hashedPassword, admin: admin } });
    });
}
exports.updateUser = updateUser;
function deleteUser(username) {
    return __awaiter(this, void 0, void 0, function* () {
        yield db.collection('utenti').deleteOne({ username: username });
    });
}
exports.deleteUser = deleteUser;
