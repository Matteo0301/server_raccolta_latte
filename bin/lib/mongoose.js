"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateHash = exports.User = exports.clear = exports.deleteUser = exports.updateUser = exports.getUsers = exports.getUser = exports.addUser = exports.db = exports.close = exports.connect = void 0;
const mongoose_1 = require("mongoose");
const logger_1 = __importDefault(require("./util/logger"));
const bcrypt_1 = __importDefault(require("bcrypt"));
const schemas_1 = require("./schemas");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return schemas_1.User; } });
let db;
exports.db = db;
function connect(CONNECTION_STRING) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, mongoose_1.set)("strictQuery", false);
        try {
            exports.db = db = yield (0, mongoose_1.connect)(CONNECTION_STRING);
        }
        catch (error) {
            logger_1.default.error("Error connecting to MongoDB: " + error);
            process.exit(1);
        }
    });
}
exports.connect = connect;
function clear() {
    return __awaiter(this, void 0, void 0, function* () {
        yield schemas_1.User.deleteMany({});
    });
}
exports.clear = clear;
function close() {
    return __awaiter(this, void 0, void 0, function* () {
        yield db.disconnect();
    });
}
exports.close = close;
function generateHash(password) {
    let salt_rounds = 10;
    if (process.env.SALT_ROUNDS) {
        salt_rounds = parseInt(process.env.SALT_ROUNDS);
    }
    return bcrypt_1.default.hashSync(password, bcrypt_1.default.genSaltSync(salt_rounds));
}
exports.generateHash = generateHash;
function addUser(username, password, admin) {
    return __awaiter(this, void 0, void 0, function* () {
        const hashedPassword = generateHash(password);
        const user = yield getUser(username);
        if (user) {
            yield deleteUser(username);
        }
        yield schemas_1.User.create({ username: username, password: hashedPassword, admin: admin });
    });
}
exports.addUser = addUser;
function getUser(username) {
    return __awaiter(this, void 0, void 0, function* () {
        const user = yield schemas_1.User.find({ username: username }).exec();
        if (user.length > 0) {
            return user[0];
        }
        return null;
    });
}
exports.getUser = getUser;
function getUsers() {
    return __awaiter(this, void 0, void 0, function* () {
        const u = yield schemas_1.User.find().exec();
        let users = [];
        u.forEach((user) => {
            users.push({ username: user.username, admin: user.admin });
        });
        return users;
    });
}
exports.getUsers = getUsers;
function updateUser(username, password, admin) {
    return __awaiter(this, void 0, void 0, function* () {
        const hashedPassword = generateHash(password);
        yield schemas_1.User.updateOne({ username: username }, { password: hashedPassword, admin: admin });
    });
}
exports.updateUser = updateUser;
function deleteUser(username) {
    return __awaiter(this, void 0, void 0, function* () {
        yield schemas_1.User.deleteOne({ username: username });
    });
}
exports.deleteUser = deleteUser;
