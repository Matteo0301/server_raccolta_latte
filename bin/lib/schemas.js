"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
const mongoose_1 = require("mongoose");
const UserSchema = new mongoose_1.Schema({
    username: String,
    password: String,
    admin: Boolean
});
const User = (0, mongoose_1.model)('User', UserSchema, 'utenti');
exports.User = User;
