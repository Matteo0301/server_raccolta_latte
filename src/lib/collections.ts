import { NextFunction, Request, Response, Router } from "express"
import { authenticateToken } from "./util/token"
import { checkAdmin, checkTokenMatchesUser, checkValidationErrors } from "./util/auth"
import { addCollection, checkCollection, deleteCollection, getCollectionByUser, getCollections } from "./mongoose"
import { body, param } from "express-validator"


const router = Router()

async function checkDates(req: Request, res: Response, next: NextFunction) {
    //console.log('req.params :>> ', req.params);
    req.start = new Date(req.params.startdate)
    req.end = new Date(req.params.enddate)
    if (isNaN(req.start.getTime()) || isNaN(req.end.getTime())) {
        res.sendStatus(400)
    } else if (req.start > req.end) {
        res.sendStatus(400)
    } else {
        next()
    }
}

router.get('/:startdate/:enddate', [
    param('startdate').notEmpty().isString().isDate().escape(),
    param('enddate').notEmpty().isString().isDate().escape(),
    checkDates,
    authenticateToken,
    checkAdmin
], async (req: Request, res: Response) => {
    const r = await getCollections(req.start, req.end)
    res.json(r)
})

router.get('/byuser/:username/:startdate/:enddate', [
    param('username').notEmpty().isString().isAlpha().escape(),
    param('startdate').notEmpty().isString().isISO8601().isDate().escape(),
    param('enddate').notEmpty().isString().isISO8601().isDate().escape(),
    checkDates,
    authenticateToken,
    checkTokenMatchesUser,
    //checkValidationErrors
], async (req: Request, res: Response) => {

    const r = await getCollectionByUser(req.params.username, req.start, req.end)
    res.json(r)
})


router.delete('/:id', [
    param('id').notEmpty().isString().isAlphanumeric().escape(),
    authenticateToken,
    checkAdmin,
    checkValidationErrors
], async (req: Request, res: Response) => {
    const id = req.params.id
    if (!await checkCollection(id)) {
        res.status(404).send()
        return
    }
    await deleteCollection(id)
    res.status(204).send()
})

router.post('/:username/:origin', [
    param('username').notEmpty().isString().isAlpha().escape(),
    param('origin').notEmpty().isString().isAlpha().escape(),
    body('quantity').notEmpty().isNumeric().escape(),
    body('quantity2').notEmpty().isNumeric().escape(),
    authenticateToken,
    checkTokenMatchesUser,
    checkValidationErrors
], async (req: Request, res: Response) => {
    const user = req.params.username
    const quantity = req.body.quantity
    const quantity2 = req.body.quantity2
    const origin = req.params.origin
    if (quantity < quantity2) {
        res.status(400).send()
        return
    }
    const date = new Date()
    await addCollection(date, quantity, quantity2, user, origin)
    res.status(201).send()
})

export default router