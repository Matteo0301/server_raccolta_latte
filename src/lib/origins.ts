import { Request, Response, Router } from "express"
import { authenticateToken } from "./util/token"
import { checkAdmin, checkValidationErrors } from "./util/auth"
import { param } from "express-validator"
import { addOrigin, deleteOrigin, getOrigins, getCollectionsByOrigin } from "./mongoose"


const router = Router()

router.get('/', [
    authenticateToken,
    checkValidationErrors
], async (req: Request, res: Response) => {
    res.json(await getOrigins())
})

router.get('/:origin', [
    param('origin').notEmpty().isString().isAlpha().escape(),
    authenticateToken,
    checkAdmin,
    checkValidationErrors
], async (req: Request, res: Response) => {
    res.json(await getCollectionsByOrigin(req.params.origin))
})


router.delete('/:origin', [
    param('origin').notEmpty().isString().isAlpha().escape(),
    authenticateToken,
    checkAdmin,
    checkValidationErrors
], async (req: Request, res: Response) => {
    if (await deleteOrigin(req.params.origin))
        res.sendStatus(201)
    else
        res.sendStatus(409)
})

router.post('/:origin/:lat/:lng', [
    param('origin').notEmpty().isString().isAlpha().escape(),
    param('lat').notEmpty().isNumeric(),
    param('lng').notEmpty().isNumeric(),
    authenticateToken,
    checkAdmin,
    checkValidationErrors
], async (req: Request, res: Response) => {
    if (await addOrigin(req.params.origin, parseFloat(req.params.lat), parseFloat(req.params.lng)))
        res.sendStatus(201)
    else
        res.sendStatus(409)
})

export default router