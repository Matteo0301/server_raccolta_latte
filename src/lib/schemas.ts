import { model, Schema } from "mongoose";

const UserSchema = new Schema({
    username: String,
    password: String,
    admin: Boolean
})

const User = model('User', UserSchema, 'utenti')

export { User }