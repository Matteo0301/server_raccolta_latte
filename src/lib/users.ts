import { Request, Response, Router } from "express";
import { authenticateUser, checkAdmin, checkValidationErrors } from "./util/auth";
import Logger from "./util/logger";
import { authenticateToken, generateAccessToken } from "./util/token";
import { addUser, deleteUser, getUser, getUsers, updateUser } from "./mongoose";
import { param, body, header } from "express-validator"
import { morganMiddleware } from "./util/morganMiddleware";

const router = Router()

router.get('/auth/:username/:password', [
    param('username').notEmpty().isString().escape(),
    param('password').notEmpty().isString(),
    checkValidationErrors,
    authenticateUser
], (req: Request, res: Response) => {
    Logger.debug('Authentication')
    Logger.debug("username " + req.user + ', admin: ' + req.admin)
    const token = { token: generateAccessToken(req.user, req.admin), admin: req.admin }
    res.json(token)
})

router.put('/', [
    body('username').notEmpty().isString().escape(),
    body('password').notEmpty().isString(),
    body('admin').notEmpty().isBoolean(),
    header('authorization').notEmpty().isString(),
    checkValidationErrors,
    authenticateToken,
    checkAdmin
], async (req: Request, res: Response) => {
    if (req.body.username === null || req.body.password === null || req.body.admin === null) {
        res.sendStatus(400)
        return
    }
    const username = req.body.username
    const password = req.body.password
    const admin = req.body.admin
    Logger.debug('Adding user ' + username + ' with password ' + password + ' and admin ' + admin)
    if (await addUser(username, password, admin))
        res.status(201).send()
    else
        res.status(409).send()
})

router.get('/', [
    authenticateToken,
    checkAdmin
], async (req: Request, res: Response) => {
    const users = await getUsers()
    const r = { users: users }
    res.json(r)
})

router.patch('/:username', [
    param('username').notEmpty().isString().escape(),
    header('authorization').notEmpty().isString(),
    checkValidationErrors,
    authenticateToken,
    checkAdmin
], async (req: Request, res: Response) => {
    if (!req.body.password && !req.body.admin && !req.body.username) {
        res.sendStatus(400)
        return
    }

    const user = await getUser(req.params.username)

    if (!user) {
        res.sendStatus(404)
        return
    }

    const password = (req.body.password !== null) ? req.body.password : user.password
    const admin = (req.body.admin !== null) ? req.body.admin : user.admin
    const newName = (req.body.password !== null) ? req.body.username : user.username
    Logger.debug('Updating user ' + req.params.username + ' with new name ' + newName + ' password ' + password + ' and admin ' + admin)
    await updateUser(req.params.username, newName, password, admin)
    res.status(204).send()
})

router.delete('/:username', [
    param('username').notEmpty().isString().escape(),
    header('authorization').notEmpty().isString(),
    checkValidationErrors,
    authenticateToken,
    checkAdmin
], async (req: Request, res: Response) => {
    const user = await getUser(req.params.username)

    if (!user) {
        res.sendStatus(404)
        return
    }

    Logger.debug('Deleting user ' + req.params.username)
    await deleteUser(req.params.username)
    res.status(204).send()
})

export default router