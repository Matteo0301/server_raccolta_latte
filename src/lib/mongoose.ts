import { set, connect as db_connect } from "mongoose"
import Logger from "./logger"
import bcrypt from 'bcrypt'
import { User } from "./schemas"
import { get } from "http"

let db: any

async function connect(CONNECTION_STRING: string) {
    set("strictQuery", false)
    try {
        /* client = new MongoClient(CONNECTION_STRING)
        await client.connect()
        db = client.db(DATABASE) */
        db = await db_connect(CONNECTION_STRING)
    } catch (error) {
        Logger.error("Error connecting to MongoDB: " + error)
        process.exit(1)
    }
}

async function close() {
    await db.disconnect()
}

function generateHash(password: string): string {
    let salt_rounds = 10
    if (process.env.SALT_ROUNDS) {
        salt_rounds = parseInt(process.env.SALT_ROUNDS)
    }
    return bcrypt.hashSync(password, bcrypt.genSaltSync(salt_rounds))
}

async function addUser(username: string, password: string, admin: boolean) {
    const hashedPassword = generateHash(password)
    const user = await getUser(username)
    if (user) {
        await deleteUser(username)
    }
    await User.create({ username: username, password: hashedPassword, admin: admin })
}

async function getUser(username: string) {
    let user = await User.find({ username: username }).exec()
    if (user.length > 0) {
        return user[0]
    }
    return null
}

async function getUsers() {
    let u = await User.find().exec()
    return u
}

async function updateUser(username: string, password: string, admin: boolean) {
    const hashedPassword = generateHash(password)
    await User.updateOne({ username: username }, { password: hashedPassword, admin: admin })
}

async function deleteUser(username: string) {
    await User.deleteOne({ username: username })
}


export { connect, close, db, addUser, getUser, getUsers, updateUser, deleteUser, User }