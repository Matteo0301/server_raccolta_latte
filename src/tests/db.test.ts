import { addUser, close, connect, getUser, getUsers, updateUser } from "../lib/mongoose";
import { User } from "../lib/schemas";


require("dotenv").config()

const CONNECTION_STRING = process.env.CONNECTION_STRING || "mongodb://user:password@mongodb:27017/raccolta_latte?authSource=raccolta_latte"
const DATABASE = process.env.DATABASE || 'raccolta_latte'
const port = process.env.PORT




describe("Users", () => {

    beforeAll(async () => {
        await connect(CONNECTION_STRING);
    })

    afterAll(async () => {
        await close()
    })

    it("Get users", async () => {
        let users = await getUsers()
        expect(users).toMatchSnapshot()
    })

    it("Add users", async () => {
        await addUser("test", "test", false)
        let users = await getUser("test")
        expect(users).not.toBeNull()
        if (users)
            expect(users.admin).toBeFalsy()
    })

    it("Update users", async () => {
        await updateUser("test", "test", true)
        let users = await getUser("test")
        expect(users).not.toBeNull()
        if (users)
            expect(users.admin).toBeTruthy()
    })

    it("Delete users", async () => {
        await User.deleteOne({ username: "test" }).exec()
        let users = await getUser("test")
        expect(users).toBeNull()
    })


})