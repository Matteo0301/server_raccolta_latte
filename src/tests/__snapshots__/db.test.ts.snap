// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Database 1`] = `
[
  {
    "_id": "6476321f4894b5bac9018a23",
    "admin": true,
    "password": "$2y$10$ZxYbM2VZdBFWRNyUDIL1DO8P.f3zUh42l4boOrZPP0//pXQFYMHRi",
    "username": "admin",
  },
]
`;

exports[`Users Get users 1`] = `
[
  {
    "_id": "6476321f4894b5bac9018a23",
    "admin": true,
    "password": "$2y$10$ZxYbM2VZdBFWRNyUDIL1DO8P.f3zUh42l4boOrZPP0//pXQFYMHRi",
    "username": "admin",
  },
]
`;
